### PR TITLE
apiserver/pkg/server: include scheme in insecure self client config

### PR DESCRIPTION
**Pull request by ericchiang**
Noticed this during bootkube development: https://github.com/kubernetes-incubator/bootkube/issues/325

In Go 1.8's `url.Parse` became more strict, and `url.Parse("127.0.0.1:8080")` now fails.

https://beta.golang.org/doc/go1.8#net_url
https://play.golang.org/p/dw_cPeotG4

Accidentally compiled bootkube with 1.8 and tracked a panic down to the loopback client config. Though we're still using the old "genericapiserver", this seems to be translated from the old one.

The actual panic we observed was

```
E0222 19:40:11.364949       5 server.go:254] Failed to create clientset: parse 127.0.0.1:8080: first path segment in URL cannot contain colon
panic: parse 127.0.0.1:8080: first path segment in URL cannot contain colon
goroutine 35 [running]:
github.com/kubernetes-incubator/bootkube/vendor/k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion.NewForConfigOrDie(0xc420728ea0, 0xc420738e30)
        /home/eric/src/github.com/kubernetes-incubator/bootkube/vendor/k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion/core_client.go:132 +0x62
github.com/kubernetes-incubator/bootkube/vendor/k8s.io/kubernetes/pkg/master.(*Config).Complete(0xc420739db8, 0x0)
        /home/eric/src/github.com/kubernetes-incubator/bootkube/vendor/k8s.io/kubernetes/pkg/master/master.go:180 +0x40a
github.com/kubernetes-incubator/bootkube/vendor/k8s.io/kubernetes/cmd/kube-apiserver/app.Run(0xc42007a3c0, 0x0, 0x0)
        /home/eric/src/github.com/kubernetes-incubator/bootkube/vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:347 +0x1e8d
github.com/kubernetes-incubator/bootkube/pkg/bootkube.(*bootkube).Run.func1(0xc4206b01e0, 0xc420164300)
        /home/eric/src/github.com/kubernetes-incubator/bootkube/pkg/bootkube/bootkube.go:124 +0x2f
created by github.com/kubernetes-incubator/bootkube/pkg/bootkube.(*bootkube).Run
        /home/eric/src/github.com/kubernetes-incubator/bootkube/pkg/bootkube/bootkube.go:124 +0xb0
```

I don't actually know if this is the correct fix or if there should be changes to `NewForConfigOrDie`. Am looking for comments more than anything.

edit: @abourget pointed out over on bootkube that the actual panic was fixed in the internalclient by https://github.com/kubernetes/kubernetes/pull/38519.

cc @deads2k @sttts @kubernetes/sig-api-machinery-pr-reviews

### CODE REVIEW

**Review by liggitt — COMMENTED** [review]
I think @sttts PR in progress fixes this

**Comment by liggitt** [comment]
Nope, it only touches the secureservinginfo one

https://github.com/kubernetes/kubernetes/pull/41882/files#diff-dc8c2c38ff02db6ef53ffc923dade8b7L148

File: staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go [hunk 1]
```diff
@@ -145,7 +145,7 @@ func (s *ServingInfo) NewSelfClientConfig(token string) (*restclient.Config, err
 		return nil, nil
 	}
 	return &restclient.Config{
-		Host: s.BindAddress,
+		Host: "http://" + s.BindAddress,
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go]: @sttts this also assumes the insecure bind address is an actual IP, not 0.0.0.0

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go]: mind adding the same check as above:

```
host, port, err := net.SplitHostPort(s.BindAddress)
if err != nil {
  // should never happen
   return nil, fmt.Errorf("invalid insecure bind address: %q", s.BindAddress)
}
if host == "0.0.0.0" {
  host = "localhost"
}
...
Host: "http://" + net.JoinHostPort(host,port),
```

**Comment by liggitt** [comment]
would like the 0.0.0.0 check, but LGTM

File: staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go (hunk 1 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go]: Or even better put it into a small helper `func (s *ServingInfo) loopbackHost() string` and use that at both places.

**Comment by ericchiang** [comment]
@liggitt @sttts does this look right? Want any tests?

**Comment by ericchiang** [comment]
@k8s-bot bazel test this

File: staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go [hunk 2]
```diff
@@ -26,28 +26,38 @@ import (
...
 
-func (s *SecureServingInfo) NewSelfClientConfig(token string) (*restclient.Config, error) {
-	if s == nil || (s.Cert == nil && len(s.SNICerts) == 0) {
-		return nil, nil
-	}
-
-	host, port, err := net.SplitHostPort(s.ServingInfo.BindAddress)
+// loopbackBindAddress determines the address to query when constructing a
+// loopback REST client.
+func loopbackBindAddress(bindAddress string) string {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go]: It's not really the BindAddress, but the address to connect to derived from the BindAddress. So `loopbackHostForBindAddress` or something like that.

**Comment by ericchiang** [comment]
Sorry, I'm an idiot and sent code that didn't compile :/ This one should work better.

File: staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go [hunk 3]
```diff
@@ -140,12 +135,35 @@ func findCA(chain []*x509.Certificate) (*x509.Certificate, error) {
 	return nil, fmt.Errorf("no certificate with CA:TRUE found in chain")
 }
 
+// loopbackHostPort returns the host and port loopback REST clients should use
+// to contact the server.
+func (s *ServingInfo) loopbackHostPort() (string, string, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go]: better 👍

**Comment by sttts** [comment]
@ericchiang please rebase

**Comment by ericchiang** [comment]
@sttts already did.

**Pull request merged**